packages/react-storage/src/components/StorageBrowser/views/hooks/usePaginate.ts

```diff
@@ -1,55 +1,70 @@
 import React from 'react';
+import { LocationData, LocationItemData } from '../../actions';
+import { isFunction } from '@aws-amplify/ui';
 
-interface UsePaginate {
+type ListItemType = LocationItemData | LocationData;
+
+interface UsePaginate<T extends ListItemType> {
   currentPage: number;
-  handlePaginateNext: (input: {
-    resultCount: number;
-    hasNextToken: boolean;
-  }) => void;
-  handlePaginatePrevious: (input?: {}) => void;
+  highestPageVisited: number;
+  onPaginate: (page: number) => void;
   handleReset: () => void;
-  range: [start: number, end: number];
+  pageItems: T[];
 }
 
-export const usePaginate = ({
-  onPaginateNext,
-  onPaginatePrevious,
-  pageSize,
-}: {
-  onPaginateNext?: () => void;
-  onPaginatePrevious?: () => void;
+interface UsePaginateProps<T extends ListItemType> {
+  hasNextToken: boolean;
+  items: T[];
+  paginateCallback?: () => void;
   pageSize: number;
-}): UsePaginate => {
+  resultCount: number;
+}
+
+export const usePaginate = <T extends ListItemType>({
+  hasNextToken,
+  items,
+  paginateCallback,
+  pageSize,
+  resultCount,
```

Comment:
Can remove resultCount from the interface, can be derived from `items.length` inside the hook

Reply:
Yes, this is a great callout. `resultCount` actually isn't used outside of usePaginate anyways, so it can be removed three times as well.